### PR TITLE
fix(ledgerstate): audit CBOR decode limits for large mainnet structures

### LINKED ISSUE #2284: ledgerstate: audit CBOR decode limits for large mainnet structures
## Summary

CBOR decode limits for very large mainnet structures need explicit local policy and tests.

## Audit Details

Finding ID: L3, LOW network/external-dependency finding.

Affected area: CBOR decoding paths, especially snapshot/import paths that can process large maps/arrays.

The audit noted that most CBOR decoding is delegated to upstream libraries. Default map-pair limits may be insufficient for very large structures such as mainnet stake maps, while dingo also has custom large-array decoding in some paths.

Impact: future upstream default changes or untested large inputs can break mainnet-scale import/sync paths or create unbounded memory exposure if limits are raised without policy.

## What Needs to Happen

1. Identify decode paths that can process large mainnet maps/arrays.
2. Document expected local limits for each path.
3. Add regression tests or generated fixtures near and above those limits.
4. Configure upstream decoders explicitly where defaults are not acceptable.
5. Keep hard caps to avoid unbounded memory use.

## Acceptance Criteria

- Large CBOR decode limits are explicit and tested.
- Tests cover both accepted large structures and rejected oversized structures.

### PR DESCRIPTION
**Pull request by cryptodj413**
Fixes #2284

Audits the manual CBOR decode paths in `ledgerstate` for unbounded
allocation/recursion on adversarial or corrupted mainnet-scale data, and
documents/enforces explicit limits where the fxamacker wrapper's built-in
10M map-pair/array-element caps don't already apply.

**Decode paths reviewed:**
- gouroboros CBOR wrapper: relies on its internal ~10M map-pair / array-element
  limits (library default, not touched here).
- `ledgerstate`'s own manual decode paths: `decodeMapEntries`,
  `cborItemSize`/`cborItemSizeDepth`, `decodeRawElements`,
  `extractAllEraBounds`/`navigateTelescope` (Conway int-keyed records), and
  the UTxO streaming decode in `utxo.go`.

**Gap closed:** `parseUTxOsStreamingWithProgress`'s on-wire map header
count wasn't bound-checked the way `decodeMapEntries` already was. Added
`checkUTxOMapEntryCount` to enforce the same limit on that path.

**Left as-is (documented, not changed):** two bare-fxamacker call sites —
block header extraction in `internal/node/load.go` and midnight's
self-written snapshot data in `midnight/indexer/indexer.go` — are fine at
library defaults since both operate on bounded/self-produced data rather
than arbitrary peer/mainnet input; noted inline and in ARCHITECTURE.md.

**Tests:** 18 new boundary-condition regression tests in
`ledgerstate/cbor_decode_test.go` (accept-at-limit / reject-over-limit for
each decode path above).

**Docs:** `ARCHITECTURE.md` updated with the decode-limits policy.
`DATABASE.md` was checked and is not applicable — no schema, blob-store,
or CBOR-offset encoding changes here.

**Not run:** the network-gated Mithril integration test
(`TestImportLedgerStateFromMithril`) and DevNet were not run for this
change — judgment call that this is a low-severity, non-consensus
doc/robustness change; happy to run either if a reviewer wants it.

---
## Summary by cubic
Audit and enforce CBOR decode limits in `ledgerstate` to prevent unbounded allocation/recursion on mainnet-scale data. Adds hard caps for UTxO map streaming (definite and indefinite) and documents that partial commits on rejection are safe and idempotent.

- **Bug Fixes**
  - UTxO maps: 10,000,000-entry cap; definite-length checked up front (`checkUTxOMapEntryCount`), indefinite-length checked per entry (`checkUTxOMapRunningEntryCount` in `parseIndefiniteUTxOMapWithProgress`) with clear errors explaining safe retry semantics.
  - Manual map walker: `decodeMapEntries` now enforces the same cap; added test-only `decodeMapEntriesLimit`.
  - Recursion/alloc guards: `cborItemSize` capped at 128 levels; HardFork telescope traversal at 16; `decodeRawElements` caps int keys at 256.
  - Docs: expanded "Local CBOR decode limits policy" in code and `ARCHITECTURE.md`; clarified safe uses of `fxamacker/cbor` defaults in `internal/node/load.go` and `midnight/indexer/indexer.go` (no behavior change).
  - Tests: boundary tests for accept-at-limit/reject-over-limit; regression proving rejected indefinite-length UTxO maps may partially commit but are idempotent and safe to retry.

<sup>Written for commit 30a78b9b20b63387cd423dd52c93784676ab7bcd. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/dingo/pull/3141?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by cryptodj413** [comment]
Fixed in e49824ce: added `checkUTxOMapRunningEntryCount` enforced per-entry in `parseIndefiniteUTxOMapWithProgress` to cap indefinite-length UTxO maps the same way `checkUTxOMapEntryCount` already caps the definite-length header count. Also narrowed the decode-limits policy comment (in `cbor_decode.go` and `ARCHITECTURE.md`) to distinguish full `cbor.Decode` calls (bounded by gouroboros's internal decode-mode cap) from manual header-reading APIs like `DecodeMapHeader`/`NewStreamDecoder`, which are not bounded by the decoder itself and need an explicit call-site check — as this fix demonstrates. New boundary tests added (`TestCheckUTxOMapRunningEntryCount`, `TestParseIndefiniteUTxOMap_AtLimitAccepted`, `TestParseIndefiniteUTxOMap_OverLimitRejected`). `go build ./...` and `go test -race ./ledgerstate/...` pass.

**Comment by cryptodj413** [comment]
Investigated in c6b15d71. The mechanics in this finding are correct — `checkUTxOMapRunningEntryCount` has no upfront header count, so batches below the cap can be committed via `importUTxOs`'s batch callback before entry `maxMapEntries+1` triggers rejection. However, this isn't a consistency bug given the existing architecture: every UTxO row goes through `CreateUtxoIfAbsent`, an idempotent insert-if-absent upsert (same pattern used by every other import phase in `ledgerstate/import.go`), and `ImportLedgerState` only calls `setCheckpoint(cfg, models.ImportPhaseUTxO)` after the phase succeeds — so a retry (resumed or from-scratch) converges to the same rows with no duplication, and the partial state is never checkpointed as a complete/usable ledger state.

Given that, I opted against both suggested fixes: a preflight count would require a second full pass over a potentially multi-GB indefinite-length UTxO-HD file just to catch a rare corrupted/adversarial case, and wrapping the whole UTxO phase in one transaction would hold a single transaction open across 10M+ rows, defeating the batched-commit design used throughout `import.go`. Confirmed the definite-length path (`checkUTxOMapEntryCount`) is unaffected — it already validates the header count before any streaming starts.

Fix: a clearer, more actionable error message at the rejection site, doc updates recording the safety rationale (`cbor_decode.go`, `ARCHITECTURE.md`), and a new regression test (`TestIndefiniteUTxOMapPartialCommitIsSafeToRetry`) proving the property against a real test database: reject at limit → two batches committed → full re-run converges to exactly the right row count with no duplicates. `go build ./...`, `go test -race ./ledgerstate/...`, and `golangci-lint` all pass.

**Review by wolf31o2 — APPROVED** [review]

File: ledgerstate/import_test.go [hunk 1]
```diff
@@ -1055,113 +1055,78 @@ func TestImportPoolsPreservesRewardAccountCredentialTag(t *testing.T) {
...
 
-// TestSynthesizeRetiredScheduledPoolsResolvesVrfKey verifies that a pool
-// present only in the imported active pool distribution (absent from the live
-// pool table) becomes resolvable via GetPool(includeInactive=true) carrying the
-// VRF key hash from the distribution, and is tombstoned with a retirement at
-// the snapshot epoch. This mirrors a pool that retired at the epoch boundary
-// but still leads the current epoch's fixed schedule, whose header VRF-key
-// binding check would otherwise fail on a Mithril-imported node.
-func TestSynthesizeRetiredScheduledPoolsResolvesVrfKey(t *testing.T) {
+// TestIndefiniteUTxOMapPartialCommitIsSafeToRetry proves the cubic-dev-ai
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply ledgerstate/import_test.go:1058]: It's fine
